crates/ra_hir/src/ty/infer.rs

```diff
@@ -483,9 +515,60 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
             Resolution::Def(def) => {
                 let typable: Option<TypableDef> = def.into();
                 let typable = typable?;
+
+                let ty = self.db.type_for_def(typable, Namespace::Values);
                 let substs = Ty::substs_from_path(self.db, &self.resolver, path, typable);
-                let ty = self.db.type_for_def(typable, Namespace::Values).subst(&substs);
+                let ty = ty.subst(&substs);
                 let ty = self.insert_type_vars(ty);
+
+                // plug the old parent_ty in
+                let plug_self_types = || -> Option<()> {
+                    let actual_def_ty = actual_def_ty?;
+
+                    if let crate::ModuleDef::Function(func) = def {
```

Comment:
I think it should be possible to do this in a simpler, more obviously correct way: just take the target type of the impl and the `actual_def_ty`, go through them in a similar way to `unify` (but not with unify!), and whenever you encounter a type parameter of the impl, record the type that is in the same place in `actual_def_ty`. This can be a simple function that takes the two types and returns a `Substs`, without any other dependencies. It doesn't actually need the whole unification machinery, because it doesn't need to unify in both directions.

Reply:
Yeah, your approach is much better and simpler than mine. Let me try to do in that way.